shinyswatch/picker/theme_picker.js

```diff
@@ -0,0 +1,50 @@
+// Get the source path of the shinyswatch-js script
+// to figure out where the htmldependency has ended up
+// e.g. src="lib/shinyswatch-js-5.3.1/bootstrap.bundle.min.js"
+// avoids having to know the shinswatch bootstrap version or lib location
+const sw_script = document.querySelector('script[src*="shinyswatch-js"');
+const subdir = sw_script.src
+  .replace(/\/bootstrap.*js$/, '')
+  .replace("shinyswatch-js", "shinyswatch-css-all");
+
+const display_warning = setTimeout(function() {
+  window.document.querySelector("#shinyswatch_picker_warning").style.display = 'block';
+}, 1000);
+
+Shiny.addCustomMessageHandler('shinyswatch-hide-warning', function(message) {
+  window.clearTimeout(display_warning);
+});
+
+Shiny.addCustomMessageHandler('shinyswatch-refresh', function(message) {
+  window.location.reload();
+});
+
+Shiny.addCustomMessageHandler('shinyswatch-pick-theme', function(theme) {
+  const base_dir = `${subdir}/${theme}`;
+
+  const sheets = ["bootswatch.min.css", "shinyswatch-ionRangeSlider.css"];
+
+  for (const sheet of sheets) {
+    const oldLink = document.querySelector(`link[data-shinyswatch-css="${sheet}"]`);
+
+    if (oldLink) {
+      if (oldLink.dataset.shinyswatchTheme !== theme) {
+        const newLink = oldLink.cloneNode(true);
+        newLink.href = newLink.href.replace(oldLink.dataset.shinyswatchTheme, theme);
+        newLink.href = newLink.href.replace(window.location.href, '');
+        newLink.dataset.shinyswatchTheme = theme;
+        oldLink.parentNode.insertBefore(newLink, oldLink.nextSibling);
+        setTimeout(() => oldLink.remove(), 200);
```

Comment:
Unfortunately I know from experience that there's no guarantee the timing on this will work out. Especially on a slow internet connection where the network request for the `newLink` takes a while more than `200ms` (in which case, you'll get a FUOC).

To do this sort of thing "properly" in shiny.js, we ended up with this https://github.com/rstudio/shiny/blob/420a2c/srcts/src/shiny/render.ts#L319-L363

Reply:
Yeah, I was okay with that -- we currently have a flash of wrongly-styled content and I'd be okay with a regular FUOC. That said, I bet I can use the `transitionend` trick here

Reply:
Is there a reason why we can't use the old approach of returning `HTMLDependency()`s to get new stylesheets? I know it's weird, but it is a clever way to take advantage of the `shiny.js` code paths to refresh `<link>`s on the same `HTMLDependency()` (which is how dynamic theming works on the R side)

Reply:
New approach in 13a665dce63c0ae2f5e2eef7fcf2958eb7273df3 uses `transitionend` on `<body>` which is _pretty much_ guaranteed by the `* { transition: all ... }` rule added by the theme picker. In case it isn't, we keep the `setTimeout()` with a longer timeout and remove the old link.

Reply:
This is much more robust now and has very much passed my threshold of _good enough for the task at hand_.